### ISSUE TITLE
Comments are downloaded twice

### ISSUE THREAD

**Issue by gabldotink** [bibanon/tubeup#252]
**Checklist**

# If you just delete all this text and post a issue it will be closed on sight.

Carefully read and work through this check list in order to prevent the most common mistakes and misuse of tubeup, put x into all relevant boxes (like this [x])

- [x] I have fully updated tubeup, youtube-dl and internetarchive along with their associated dependencies.
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the issues (closed or open) for similar bug reports including closed ones
- [x] I am not submitting a bug report about S3 timeouts (we have no control over Internet Archive outages or per-user throttling)
- [x] I've checked the Twitter account of Internet Archive (@internetarchive) for any announced outages or planned infrastructure maintenance that would affect performance of uploads.
- [x] I have properly indented [with triple backticks before and after the console output](https://guides.github.com/pdfs/markdown-cheatsheet-online.pdf) full terminal terminal output from the line where the command was run to where I was returned to command prompt, and am not trying to obscure item identifiers or URLs used in the creation of the bug (we need this to recreate the bug or investigate what happened).

**Dependency versions**

Please provide version information from core dependencies:

`ia --version`
`3.0.2`

**Expected behavior**
A video's comments should only be downloaded _once_, and the Internet Archive should be checked for an existing item before that. More on what actually happens below.

**Additional context**
(Example console output is [here](https://github.com/bibanon/tubeup/files/9865593/log.txt) as a text file due to its length. Most of the log is not relevant. Line numbers are provided in this explanation. The video is [this one](https://www.youtube.com/watch?v=RHKw_pTPs08), and I'm running the program using Termux on Android 12. I have tested it on other devices and different videos, and the same thing happens, but the log and video are for those specifically. It is not a platform-specific problem) The log does not contain the file uploads because it didn't get `tee`d to the file correctly; however, the files were uploaded [here](https://archive.org/details/youtube-RHKw_pTPs08) so you can check them.

Currently, the option `--get-comments` seems to download a video's comments twice. If there's a reason for this, please let me know, but I can't currently think of one.

In addition, the first time comments are downloaded comes _before_ checking if the item already exists on the Internet Archive. I'm guessing this is probably because the first download is a bug. I'm not submitting this as a separate issue because it appears to stem directly from this one. The second comments download works without issue after the check for existing items occurs.

**Comment by vxbinaca** [comment]
Look at the proper indenting and not deleting the submission. Props to the submitter.

The first checked box, you didn't do it. Upgrade `tubeup` then tell me about the existing item thing. Go ahead try it.

Use pip to upgrade `Tubeup`. I'll wait.

**Comment by vxbinaca** [comment]
@mrpapersonic Does it check for the item before downloading comments? I can't see where the comments are journaled in his item, and he didn't include the text output because mobile.

I need the text output or `--debug` to see whats happening. It would help greatly if you got a computer to do this with.

**Comment by gabldotink** [comment]
`tubeup` is fully upgraded. `pip install --upgrade tubeup` simply prints that requirements are satisfied, then exits. It doesn't show up when running `pip list --outdated`.

As for the debug logging, the exact command I used was `tubeup https://www.youtube.com/watch?v=RHKw_pTPs08 --debug --get-comments`.

I apologize for using my phone for this; it was the most convenient device for me at the time. I will try to get a log from my PC ASAP.

**Comment by vxbinaca** [comment]
I need the terminal output in debug to see whats going on. That command is unhelpful because I can't simulate it, because you own the item. 

So lemme try to go grab something to test.

Comments are more trouble than they're worth.

**Comment by gabldotink** [comment]
Oops! I made a mistake. I tried to use `sed` to remove ANSI color codes and carriage returns, but it ended up deleting a lot of essential parts of the log. I sincerely apologize for this. This log has the color codes, but the text is still readable enough, and using a tool like `cat` or `bat` will correctly display the log with color. Correct log: [log2.txt](https://github.com/bibanon/tubeup/files/9865768/log2.txt)

I didn't intend to use mobile as an "excuse" for a broken log, although the problem was something different anyway. And Termux is more than capable of running applications correctly, especially since applications (such as the Python I used) are made sure to compile and execute in the correct way when used. I've done several things that are standard tasks in Linux with no issue.

I'd have to disagree that comments are "more trouble than they're worth." The entire point of archiving the internet this way is to preserve the opinions, thoughts, jokes, culture of entire generations, hopefully forever. Comments are 100% worth it.

**Comment by vxbinaca** [comment]
Yup you're right. It does grab comments before checking for an upload. Consider features like getting comments to be beta. The reason we don't have comments enabled by default, despite incessant requests to enable it by default, is this exact issue.

`tubeup --debug --get-comments https://www.youtube.com/watch\?v\=eZfwYVgbuPM`

```
tubeup --debug --get-comments https://www.youtube.com/watch\?v\=eZfwYVgbuPM
[DEBUG] 2022-10-25 23:32:51,914 - tubeup.TubeUp - DEBUG - [debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[DEBUG] 2022-10-25 23:32:51,914 - tubeup.TubeUp - DEBUG - [debug] yt-dlp version 2022.10.04 [4e0511f] (pip) API
[DEBUG] 2022-10-25 23:32:51,920 - tubeup.TubeUp - DEBUG - [debug] Python 3.10.6 (CPython 64bit) - Linux-5.15.0-53-generic-x86_64-with-glibc2.35 (glibc 2.35)
[DEBUG] 2022-10-25 23:32:51,920 - tubeup.TubeUp - DEBUG - [debug] Checking exe version: ffmpeg -bsfs
[DEBUG] 2022-10-25 23:32:51,976 - tubeup.TubeUp - DEBUG - [debug] Checking exe version: ffprobe -bsfs
[DEBUG] 2022-10-25 23:32:52,028 - tubeup.TubeUp - DEBUG - [debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2
[DEBUG] 2022-10-25 23:32:52,029 - tubeup.TubeUp - DEBUG - [debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2022.09.24, mutagen-1.45.1, secretstorage-3.3.1, sqlite3-2.6.0, websockets-10.3
[DEBUG] 2022-10-25 23:32:52,099 - tubeup.TubeUp - DEBUG - [debug] Proxy map: {}
[DEBUG] 2022-10-25 23:32:52,141 - tubeup.TubeUp - DEBUG - [debug] Loaded 1690 extractors
[DEBUG] 2022-10-25 23:32:52,162 - tubeup.TubeUp - DEBUG - [debug] [youtube] Extracting URL: https://www.youtube.com/watch?v=eZfwYVgbuPM
[DEBUG] 2022-10-25 23:32:52,162 - tubeup.TubeUp - DEBUG - [youtube] eZfwYVgbuPM: Downloading webpage
[DEBUG] 2022-10-25 23:32:53,029 - tubeup.TubeUp - DEBUG - [youtube] eZfwYVgbuPM: Downloading android player API JSON
[DEBUG] 2022-10-25 23:32:53,574 - tubeup.TubeUp - DEBUG - [debug] Sort order given by extractor: quality, res, fps, hdr:12, source, vcodec:vp9.2, channels, acodec, lang, proto
[DEBUG] 2022-10-25 23:32:53,574 - tubeup.TubeUp - DEBUG - [debug] Formats sorted by: hasvid, ie_pref, quality, res, fps, hdr:12(7), source, vcodec:vp9.2(10), channels, acodec, lang, proto, filesize, fs_approx, tbr, vbr, abr, asr, vext, aext, hasaud, id
[DEBUG] 2022-10-25 23:32:53,679 - tubeup.TubeUp - DEBUG - [youtube] Downloading comment section API JSON
[DEBUG] 2022-10-25 23:32:54,491 - tubeup.TubeUp - DEBUG - [youtube] Downloading ~285 comments
[DEBUG] 2022-10-25 23:32:54,491 - tubeup.TubeUp - DEBUG - [youtube] Sorting comments by newest first
[DEBUG] 2022-10-25 23:32:54,492 - tubeup.TubeUp - DEBUG - [youtube] Downloading comment API JSON page 1 (0/285)
[DEBUG] 2022-10-25 23:32:55,142 - tubeup.TubeUp - DEBUG - [youtube] Downloading comment API JSON page 2 (20/285)
[DEBUG] 2022-10-25 23:32:55,800 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 1 (36/285)
[DEBUG] 2022-10-25 23:32:56,402 - tubeup.TubeUp - DEBUG - [youtube] Downloading comment API JSON page 3 (44/285)
[DEBUG] 2022-10-25 23:32:57,027 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 1 (47/285)
[DEBUG] 2022-10-25 23:32:57,573 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 2 (61/285)
[DEBUG] 2022-10-25 23:32:58,179 - tubeup.TubeUp - DEBUG - [youtube] Downloading comment API JSON page 4 (67/285)
[DEBUG] 2022-10-25 23:32:58,825 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 1 (69/285)
[DEBUG] 2022-10-25 23:32:59,445 - tubeup.TubeUp - DEBUG - [youtube]        Downloading comment replies API JSON page 1 (77/285)
[DEBUG] 2022-10-25 23:33:00,021 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 2 (82/285)
[DEBUG] 2022-10-25 23:33:00,578 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 3 (90/285)
[DEBUG] 2022-10-25 23:33:01,156 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 4 (99/285)
[DEBUG] 2022-10-25 23:33:01,690 - tubeup.TubeUp - DEBUG - [youtube] Downloading comment API JSON page 5 (101/285)
[DEBUG] 2022-10-25 23:33:02,304 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 1 (103/285)
[DEBUG] 2022-10-25 23:33:02,919 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 2 (113/285)
[DEBUG] 2022-10-25 23:33:03,505 - tubeup.TubeUp - DEBUG - [youtube]        Downloading comment replies API JSON page 1 (123/285)
[DEBUG] 2022-10-25 23:33:04,150 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 3 (142/285)
[DEBUG] 2022-10-25 23:33:04,698 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 4 (145/285)
[DEBUG] 2022-10-25 23:33:05,278 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 5 (150/285)
[DEBUG] 2022-10-25 23:33:05,843 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 6 (160/285)
[DEBUG] 2022-10-25 23:33:06,396 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 7 (163/285)
[DEBUG] 2022-10-25 23:33:06,923 - tubeup.TubeUp - DEBUG - [youtube] Downloading comment API JSON page 6 (168/285)
[DEBUG] 2022-10-25 23:33:07,619 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 1 (171/285)
[DEBUG] 2022-10-25 23:33:08,135 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 2 (174/285)
[DEBUG] 2022-10-25 23:33:08,724 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 3 (181/285)
[DEBUG] 2022-10-25 23:33:09,259 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 4 (184/285)
[DEBUG] 2022-10-25 23:33:09,825 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 5 (191/285)
[DEBUG] 2022-10-25 23:33:10,382 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 6 (194/285)
[DEBUG] 2022-10-25 23:33:10,947 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 7 (198/285)
[DEBUG] 2022-10-25 23:33:11,562 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 8 (206/285)
[DEBUG] 2022-10-25 23:33:12,108 - tubeup.TubeUp - DEBUG - [youtube]        Downloading comment replies API JSON page 1 (216/285)
[DEBUG] 2022-10-25 23:33:12,981 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 9 (235/285)
[DEBUG] 2022-10-25 23:33:13,541 - tubeup.TubeUp - DEBUG - [youtube] Downloading comment API JSON page 7 (239/285)
[DEBUG] 2022-10-25 23:33:14,146 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 1 (241/285)
[DEBUG] 2022-10-25 23:33:14,741 - tubeup.TubeUp - DEBUG - [youtube]        Downloading comment replies API JSON page 1 (251/285)
[DEBUG] 2022-10-25 23:33:15,301 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 2 (255/285)
[DEBUG] 2022-10-25 23:33:15,860 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 3 (264/285)
[DEBUG] 2022-10-25 23:33:16,414 - tubeup.TubeUp - DEBUG - [youtube]     Downloading comment API JSON reply thread 4 (267/285)
[DEBUG] 2022-10-25 23:33:16,989 - tubeup.TubeUp - DEBUG - [youtube] Extracted 276 comments
[DEBUG] 2022-10-25 23:33:17,012 - tubeup.TubeUp - DEBUG - [debug] Default format spec: bestvideo*+bestaudio/best
[DEBUG] 2022-10-25 23:33:17,015 - internetarchive.session - DEBUG - no metadata provided for "youtube-eZfwYVgbuPM", retrieving now.
[DEBUG] 2022-10-25 23:33:17,018 - urllib3.connectionpool - DEBUG - Starting new HTTPS connection (1): archive.org:443
[DEBUG] 2022-10-25 23:33:17,582 - urllib3.connectionpool - DEBUG - https://archive.org:443 "GET /metadata/youtube-eZfwYVgbuPM HTTP/1.1" 200 None

:: Item already exists. Not downloading.
Title: biden death grips
Video URL: https://www.youtube.com/watch?v=eZfwYVgbuPM

[DEBUG] 2022-10-25 23:33:17,583 - tubeup.TubeUp - DEBUG - Basenames obtained from url (https://www.youtube.com/watch?v=eZfwYVgbuPM): set()
```

**Comment by gabldotink** [comment]
Okay, I forgot to add line numbers in my initial issue report, but the issue makes itself obvious very quickly. If you do still need them, I'll look through and grab them.

The `info.json` for the item I uploaded is attached here: [comments.json.txt](https://github.com/bibanon/tubeup/files/9865813/comments.json.txt). The comments begin considerably down line `7441` (it is a very long line; do a text search for "I love how he uses a unique different style` to get to the first comment.

Comments being a beta feature means that this is an important issue, so that's good.

**Comment by vxbinaca** [comment]
> The entire point of archiving the internet this way is to preserve the opinions, thoughts, jokes, culture of entire generations, hopefully forever. Comments are 100% worth it.

Comments aren't a element of the video but of the website. The Wayback machine is a web archive. Tubeup is not it is a video archiver. Comments are not the same as subtitles, metadata, thumbnails and other video elements. They're 0% worth it for this exact reason. They take too long to parse, Youtube will hammer down on scrapes of it's site and likely ban your IP address, as enumerating comments isn't browser-like behavior but that of a bot.

Did the video creator create all those comments? No? Then it's not related to the video. 

Where does it stop with the mobile phone ripping groups? Shall we grab the site too? How about the webpage around the video. It's apart of the experience, right? How about we do full rips of everything thats rendered in the right side list of a single video. We get all those and their comments too.

Tubeup gets single videos at a time handed to it by yt-dlp. Comments are a rip at your own risk thing. IA doesn't have a way to process it out of the JSON and likely won't. And it will *never* be enabled by default for this issues exact reason.

**Comment by gabldotink** [comment]
**Please read this, because it is important.**
> Shall we grab the site too? How about the webpage around the video. It's apart of the experience, right? How about we do full rips of everything thats rendered in the right side list of a single video. We get all those and their comments too.

Exactly correct. That's why the Wayback Machine does this in the first place. They even have an [entire IA-staff-sanctioned collection for it](https://archive.org/details/youtubecrawl).
> Comments aren't a element of the video but of the website. The Wayback machine is a web archive. Tubeup is not it is a video archiver. Comments are not the same as subtitles, metadata, thumbnails and other video elements.
1. Your first statement is debatable (but a valid argument), but it is just as important to archive in any case.
2. The Wayback Machine is not perfect; it has a hard time downloading dynamic website content for many websites, including (read: especially sites like) YouTube. This includes comments. (see [this snapshot](https://web.archive.org/web/20210116231049/https://www.youtube.com/watch?v=RHKw_pTPs08&gl=US&hl=en) of the video I archived.) This very much includes comments, because for a video with a lot you could manually open comment threads for _hours_ and still not get all of them. `tubeup` is made to be a tool for convenience. All of the things it does can be done manually, but it is used in spite of that because it is simple, quick and easy to use. And `--get-comments` is massively helpful in this task. Beta feature? That's fine! Important feature? Incredibly so.
> They're 0% worth it for this exact reason. They take too long to parse, Youtube will hammer down on scrapes of it's site and likely ban your IP address, as enumerating comments isn't browser-like behavior but that of a bot.
1. They do not take too long to parse. They take a long time to parse, but that is not too long, especially when `tubeup` does all the work for you in the background.
2. YouTube doesn't do IP bans. YouTube _knows_ that YouTube is used in a bot-like way by many things; if they didn't know that, they wouldn't have an API to fetch comments and metadata in the first place. And YouTube is developed and expected to have literal _petabytes_ of data served every _day_. Google, its parent company, processes _exabytes_ daily across all of their services.
> Did the video creator create all those comments? No? Then it's not related to the video.

Entirely incorrect. Comments on a video preserve so, so much culture and even just important information for information's sake. They are 100% related to the video, or else they would be in a different place; and as previously mentioned, the Wayback Machine _does not get these_. And having them in a huge JSON file does not at all harm a casual Archive user's experience. In fact, most people will probably never touch that JSON file, if they even come across the item at all. They'll go, "huh? what's a json?" and then not worry about it. But for future people, that data is important, because without the Archive, it would be _gone_. Permanently. And that's why we need it. It's like saying that most Egyptian papyri is unimportant and/or uninteresting because it doesn't contain reports of gods and epic tales and kings and all that. Of course it's important. It's a piece of human history, no matter how small it is.
> Comments are a rip at your own risk thing.

There is no risk, and not even any real unintended behavior at the end of the day; comments are written to the info JSON once and only once. But it is an unintended and unneeded behavior of the application. That's why I submitted the issue. Doesn't even matter if it's never resolved.
> IA doesn't have a way to process it out of the JSON and likely won't.

And there are specific tools with the express purpose of processing JSON. `jq` comes to mind off the top of my head, and `jq` is used as part of tons and tons software projects for all types of purposes. The Internet Archive doesn't need a tool to process it. And the comment JSON is so simple, it could even be parsed by _hand_ if it needed to be.
> And it will _never_ be enabled by default for this issues exact reason.

You appear to be one of the main maintainers of this project. A bit of constructive criticism? This is a _horrible_ mindset to have, especially as the one who gets a huge say in this project. If there becomes a reason to enable it by default, _it should be enabled by default_. If it's not ready to be like that at this stage, then it's not, and that's good. But saying an incredibly important feature will _never_ have a certain quality is not great.
> Where does it stop with the mobile phone ripping groups?

I'll be honest, I have no idea what this means. _Mobile phones are a viable way to do work_. They're just computers when push comes to shove.

With all due respect &mdash; because you do deserve a lot of respect for all the hard (and often thankless) work you do &mdash; you are so wrong about so many things. I'm wrong about a lot of stuff too. I might even objectively be wrong on some of this. But this is an important conversation to have. I'm especially worried because you've closed literally hundreds of issues out of hand because you don't believe in it. Even when they're great ideas. It's your job to discuss with the community what should and shouldn't be a part of `tubeup`. In the end, it's your call, but a lot of thought needs to be put into that decision.

**Comment by vxbinaca** [comment]
>You appear to be one of the main maintainers of this project.

Correct. Fork it if you don't like the direction it's taking and the scope is too small or if you don't like my keeping the scope of tubeup small. I've closed hundreds of issues from people who don't understand how tubeup works (they want extractor fixes that we don't do because it's not our project to download video), who can't submit proper bug reports, who want Tubeup to do things that yt-dlp/youtube-dl does if they'd just read the documentation on those. who can't read the front page that tells them how to do what they want.

All of these features are weight, and this weight is increasingly for a niche subset of users with specific demands. This weight needs to be maintained, like the comments flag thats now broken because of the increasing complexity of the script that maintainers whos efforts though valiant cannot comprehend increasing complexity of this tool.

Might I remind you that IA staff have used this tool. Other groups who've pushed literal millions of videos onto the site. 

I don't turn down bad PRs and consider requests in issues. Shall I just accept all ideas even bad ones? Am I a tyrant for asking for stability of a tool I and others use daily? No.

Fork it if you don't like it. Please. This issue is a legitmate bug and should be fixed. We're not turning on comments by default. Most people want to quickly grab a video. Comments are a website element. We don't do the web we do videos. Fork it if you don't like this position. It's not destruction of history. You're free to go beat me to the punch on all my targets (my scripts are up for the public to see) and take over for my archival targets if you want comments.

**Comment by vxbinaca** [comment]
>They do not take too long to parse. They take a long time to parse, but that is not too long, especially when tubeup does all the work for you in the background.

This one I want to address because I tested this this morning.

Small video thats a quick rip, took 45 second to parse a few hundred comments. A HOT video with 20,000 comments took like 10-15 minutes. The videos would have gone up both in under a minute or just over a minute respectively. Thats an eternity for a tool thats designed to quickly grab video before it's removed. Use the comments flag yourself, or better yet, go make a project that scrapes nothing but YT comments and put them into a separate item/collection.

I mean we don't do the ads. But it's part of the experience, right? Am I ruining history by not getting ads now? What about the ideas and jokes and debates of ADVERTISEMENTS? Am I depriving future generation of having to sit through 2-9 ads per video?

But back on topic, this bug should be fixed.

**Comment by gabldotink** [comment]
OK, I'm sorry. I had a strong opinion that didn't align with yours, and I'm sorry. I fell into the whole "someone is wrong on the internet" thing when I shouldn't have.

I appreciate your hard work and your desire to keep things simple, I really do. I was just not having a good time when writing last night. It's no excuse.

**Comment by vxbinaca** [comment]
Look I'll get the bug fixed but your obsession with youtube comments, ain't Tubeups deal. It negatively impacts video ripping and is a time sink. Figure that out on your own, My advice, is use yt-dlp and have it just journal the JSON with comments and the file named as the video ID of the youtube video. You could with a VPS, shove millions of comment snapshots to IA. Talk to them about that.

But turning comment ingestion on by default impacts everyone else for a narrow band of mobile phone users.  If I wanted to be mean to comment-stans I'd have not accepted the `--comments` flag in the first place.

Do comments on your own but multiple devs don't want them on by default.

**Comment by gabldotink** [comment]
I respect that.
